packages/mobile/src/services/offline-downloader/offline-downloader.ts

```diff
@@ -0,0 +1,102 @@
+import path from 'path'
+
+import type { Track, UserMetadata } from '@audius/common'
+import {
+  encodeHashId,
+  accountSelectors,
+  cacheTracksSelectors
+} from '@audius/common'
+import { uniq } from 'lodash'
+import RNFS, { exists } from 'react-native-fs'
+
+import { store } from 'app/store'
+
+import { apiClient } from '../audius-api-client'
+
+import {
+  getLocalAudioPath,
+  getLocalCoverArtPath,
+  getLocalTrackJsonPath
+} from './offline-storage'
+const { getUserId } = accountSelectors
+const { getTrack } = cacheTracksSelectors
+
+// Main entrypoint - perform all steps required to complete a download
+export const downloadTrack = async (trackId: number, collection: string) => {
+  const state = store.getState()
+  const track = getTrack(state, { id: trackId })
+  if (!track) return false
+
+  const coverArtUri = getCoverArtUri(track)
+  const [coverArtDirectory, coverArtFileName] = getLocalCoverArtPath(track)
+  if (coverArtUri && coverArtDirectory && coverArtFileName) {
+    await downloadIfNotExists(coverArtUri, coverArtDirectory, coverArtFileName)
+  }
+  await tryDownloadTrackFromEachCreatorNode(track)
+  await writeTrackJson(track, collection)
```

Comment:
What do we do if either of these steps fail?

Reply:
Really, we should check if all of the assets exist before writing `downloaded_time` to the track json. There are more steps that need to happen here, so I haven't gotten to that yet, but it's 100% something we need to worry about. I'll add a todo for the next round.